### PR TITLE
Fix #1461 -- Validate required fields

### LINKED ISSUE #1461: Questionnaire Validation Is Not Working for Required Fields
### Steps to reproduce the error
1. Upload questionaire without any of the required fields
2. Questionnaire uploads succesfully.

### Expected behavior
I expect to get an error about missing party_name or location_type. 

While the form uploads successfully, the record will not submit in GeoODK.

### PR DESCRIPTION
**Pull request by oliverroick**
#### Why I made this change

The Cadasta platform has a set of required fields that need to be provided whenever a record is created. Using XLSForms, it is currently possible to create data schemas that omit these fields, which in turn makes it possible to submit data from ODK resulting in an error. 

This PR provides a fix for the problem by validating required fields during the upload of an XLSForm. 

#### Description of the change

The PR adds:

- The function `validate_required` to `questionnaires.validators`, which takes in a JSON representation of all fields defined in a questionnaire. The JSON representation is usually the [result of `pyxform.xls2json.parse_file_to_json`](https://github.com/Cadasta/cadasta-platform/blob/a418c1e38082a80d5d10a0b122730c620a7677d3/cadasta/questionnaires/managers.py#L222). 
- A set of helper functions, which are called from `validate_required`. I provided inline comments about the functionality, which I hope are sufficient to understand how the code works. 
- Changes to existing questionnaires so they contain the required fields. 

You can probably tell from my excessive use of `map`, `filter` and `reduce` that I have been reading up on Python's built-in function. The proposed solution might be a bit too clever If that's the case I'm happy to discuss alternatives. 

##### Validation rules
 
A valid questionnaire meets the following conditions.

The following fields must be defined, of the correct type and `required` must be `yes` in the form:

Field name      | Type
--------------- | -------------
`location_type` | `select one`
`party_name`    | `text`
`party_type`    | `select one`
`tenure_type`   | `select one`

At least one of the following fields must be present. All fields that are present must be of the correct type and `required` must be `yes` in the form:

Field name          | Type
------------------- | -------------
`location_geoshape` | `geoshape`
`location_geotrace` | `geotrace`
`location_geometry` | `geopoint`

#### How someone else can test the change

There are a lot of possible combinations that make questionnaire upload or raise an error. I suggest to start with the _Urban Informal Settlements Form_ that you can download from the site and test the following cases: 

- Upload the form as is. You should see three errors because the geometry fields are not required. 
- Make the three fields required='yes'. The form should upload successfully. 
- Remove the field `location_geometry`. The form should upload successfully. 
- Remove the fields `location_geoshape` and `location_geotrace`. You should see the error "Please provide at least one geometry field."
- Change the field `party_name` to `date`. You should see one additional error "Field party_name must be of type text.".
- Remove required='yes from the field `party_type`. You should see an additional error  "Field party_type must be required."

### When should this PR be merged

When accepted and tested. 

### Risks

The rules are strict; if the form doesn't comply with the rules outlined above, you can't upload the form. 

By its nature, everything related to questionnaires is risky business. There's a chance that I did not consider an uncommon case currently used by one of our partners. This won't affect any existing forms only those that you want to upload. 

I suggest we test this at least against forms, which are currently in the making. 

### Follow-up actions

- [ ] To keep the PR small(ish), I did not apply the changes to the API, which should be done at a later stage. I'll open a ticket and schedule this for one of the following sprints. 
- [ ] The example forms linked from the website should be updated so they can be uploaded.
- [ ] Add and implement test cases for the functional test suite to test the UI (error messages).

#### General

**Is this PR explained thoroughly?** All code changes must be accounted for in the PR description. 

- [ ] Review 1
- [x] Review 2

**Is the PR labeled correctly?** It should have the `migration` label if a new migration is added. 

- [ ] Review 1
- [x] Review 2

**Is the risk level assessment sufficient?** The risks section should contain all risks that might be introduced with the PR and which actions we need to take to mitigate these risks. Possible risks are database migrations, new libraries that need to be installed or changes to deployment scripts.

- [ ] Review 1
- [x] Review 2

#### Functionality

**Are all requirements met?** Compare implemented functionality with the requirements specification.

- [ ] Review 1
- [x] Review 2

**Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled. 

- [ ] Review 1
- [x] Review 2

#### Code

**Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.

- [ ] Review 1
- [x] Review 2

**Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests. 

- [ ] Review 1
- [x] Review 2

**Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/). 

- [ ] Review 1
- [x] Review 2

**Is the code documented sufficiently?** Large and complex classes, functions or methods must be annotated with comments following our [code-style guidelines](https://devwiki.corp.cadasta.org/Contributing%20Style%20Guide#documentation-and-comments).

- [ ] Review 1
- [x] Review 2

**Has the scalability of this change been evaluated?**

- [ ] Review 1
- [x] Review 2

**Is there a maintenance plan in place?**

- [ ] Review 1
- [x] Review 2

#### Tests

**Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components. 

- [ ] Review 1
- [x] Review 2

**If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.

- [ ] Review 1
- [x] Review 2

**If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

- [ ] Review 1
- [x] Review 2

#### Security

**Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**

- [ ] Review 1
- [x] Review 2

**Are all UI and API inputs run through forms or serializers?** 

- [ ] Review 1
- [x] Review 2

**Are all external inputs validated and sanitized appropriately?**

- [ ] Review 1
- [x] Review 2

**Does all branching logic have a default case?**

- [ ] Review 1
- [x] Review 2

**Does this solution handle outliers and edge cases gracefully?**

- [ ] Review 1
- [x] Review 2

**Are all external communications secured and restricted to SSL?**

- [ ] Review 1
- [x] Review 2

#### Documentation

**Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes must be documented in the [Cadasta Platform Documentation](https://github.com/Cadasta/cadasta-docs).

- [ ] Review 1
- [x] Review 2

**Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented in the [API docs](https://github.com/Cadasta/api-docs).

- [ ] Review 1
- [x] Review 2

**Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki. 

- [ ] Review 1
- [x] Review 2

### CODE REVIEW

**Comment by seav** [comment]
Note: The Travis PR build is failing because it is referencing the updated functional test that assumes that PR #1985 has already been merged. To be able to merge this PR, PR #1985 needs to be merged first so that the actual error message would match what is expected by the functional test.

File: cadasta/questionnaires/validators.py [hunk 1]
```diff
@@ -186,3 +188,139 @@ def validate_questionnaire(json):
...
+    Args:
+        field: string containing the field name
+
+    Returns:
+        True if the field name is in geometry_fields
+    """
+    return field.get('name') in geometry_fields.keys()
+
+
+def flatten(fields):
```
[the remark this answers is not in the document]
Comment (oliverroick) [reply cadasta/questionnaires/validators.py]: Renaming to `map_fields`

File: cadasta/questionnaires/validators.py [hunk 2]
```diff
@@ -186,3 +188,139 @@ def validate_questionnaire(json):
...
+    # joining both error lists
+    errors = list(required_errors) + list(geometry_errors)
+
+    # One geometry must be defined, so we're checking that here.
+    if not len(geometries) > 0:
+        errors.append(_('Please provide at least one geometry field.'))
+
+    errors = list(filter(None, errors))
+    if errors:
+        raise InvalidQuestionnaire(errors)
```
[the remark this answers is not in the document]
Comment (oliverroick) [reply cadasta/questionnaires/validators.py]: Ok, ok convinced. [Looks like it's an antipattern anyway](https://docs.quantifiedcode.com/python-anti-patterns/readability/using_map_or_filter_where_list_comprehension_is_possible.html). 

I changed the code so we're listcomps throughout.

File: cadasta/questionnaires/managers.py [hunk 3]
```diff
@@ -227,6 +227,7 @@ def create_from_form(self, xls_form=None, original_file=None,
                     raise InvalidQuestionnaire([
```
[the remark this answers is not in the document]
Comment (oliverroick) [reply cadasta/questionnaires/managers.py]: ✅

File: cadasta/questionnaires/validators.py [hunk 4]
```diff
@@ -287,32 +258,37 @@ def validate_field(field_def, available_fields, field):
...
 
     # Validating all geometry fields. This is a separate action because only
     # one of the three possible must be defined. We're basically just checking
     # if the geometries that are defined are ok.
     _validate_geometries = partial(validate_field,
                                    geometry_fields,
                                    required_available)
-    geometry_errors = map(_validate_geometries, list(geometries.keys()))
+    geometry_errors = [_validate_geometries(g)
+                       for g in list(geometries.keys())]
```
Comment (alukach) [inline cadasta/questionnaires/validators.py]: Okay, one last tiny nitpick: these `for g in list(geometries.keys())` statements can be written more simply as `for g in geometries` as a) `.keys()` is iterable, no need to cast it to a `list`; and b) `.keys()` is implied (you could argue that using `x in foo.keys()` is more explicit than `x in foo`)

Comment (oliverroick) [reply cadasta/questionnaires/validators.py]: Good points. I committed a clean-up

**Comment by oliverroick** [comment]
@seav I added a commit that adds the `party_type` choices validation.

**Pull request merged**